src/Nest/CommonAbstractions/Infer/IndexName/IndexNameResolver.cs

```diff
@@ -15,26 +16,28 @@ public IndexNameResolver(IConnectionSettingsValues connectionSettings)
 
 		public string Resolve(IndexName i)
 		{
-			if (i == null) return this.Resolve((Type)null);
-			return i.Name ?? this.Resolve(i.Type);
+			if (i == null || string.IsNullOrEmpty(i.Name))
+				return this.Resolve((Type)null);
+			if (char.IsUpper(i.Name[0]))
+				throw new ResolveException($"Index names cannot begin with an uppercase character: {i.Name}");
+			return i.Name;
 		}
 
 		public string Resolve(Type type)
 		{
+			var indexName = this._connectionSettings.DefaultIndex;
 			var defaultIndices = this._connectionSettings.DefaultIndices;
-
-			if (defaultIndices == null)
-				return this._connectionSettings.DefaultIndex;
-
-			if (type == null)
-				return this._connectionSettings.DefaultIndex;
-
-			string value;
-			if (defaultIndices.TryGetValue(type, out value) && !string.IsNullOrWhiteSpace(value))
-				return value;
-			return this._connectionSettings.DefaultIndex;
+			if (defaultIndices != null && type != null)
+			{
+				string value;
+				if (defaultIndices.TryGetValue(type, out value) && !string.IsNullOrEmpty(value))
+					indexName = value;
+			}
+			if (string.IsNullOrWhiteSpace(indexName))
+				throw new ResolveException($"Could not resolve an index name for the given type and no default index is set.");
+			if (char.IsUpper(indexName[0]))
```

Comment:
Can we move these lines and the ones around line 19 into its own method seeing they do almost the same thing